software/contrib/master_clock.py

```diff
@@ -0,0 +1,340 @@
+labels: clock, divider
+
+A master clock and clock divider. Each output sends a +5V trigger/gate at different divisions of the master clock, or randomly if condigured with a division of zero.
+Pulse width (gate/trigger duration) is configurable up to a maximum of 50% of the pulse width of output 1.
+
+All configuration (BPM, Pulse Width, output clock divisions) is automatically saved, then loaded when the module is restarted.
+
+For wonky/more interesting clock patterns try these:
+- Reset to step 1 using a gate into the digital input, or by using an odd value for the maximum division
+- Vary BPM by sending CV into the analog input
+- Set the division to zero for an output, this will cause the output to randomly go from high (+5V) to low (0V)
+
+Demo video: TBC
+
+digital_in: (optional) Reset step count on rising edge
+analog_in: (optional) Adjust BPM
+
+knob_1: Screen 2: Adjust BPM. Screen 3: Select output to edit 
+knob_2: Screen 2: Adjust Pulse width. Screen 3: Adjust division of selected output 
+
+button_1: Start / Stop
+button_2: Short Press (<500ms): Cycle through screens. Long Press (>500ms): Enter config mode
+
+Defaults:
+output_1: clock / 1
+output_2: clock / 2
+output_3: clock / 4
+output_4: clock / 8
+output_5: clock / 16
+output_6: clock / 32
+
+Known Issues:
+- If playback is restarted while screen 2 is in config mode, playback will be slightly irratic
+
+'''
+
+class MasterClockInner(EuroPiScript):
+    def __init__(self):
+        # Overclock the Pico for improved performance.
+        machine.freq(250_000_000)
+        self.step = 1
+        self.completedCycles = 0
+        self.running = True
+        self.CLOCKS_PER_QUARTER_NOTE = 4
+        self.MIN_BPM = 20  # Successfully calibrated to >= 20 and <= 240 BPM
+        self.MAX_BPM = 240
+        self.MIN_PULSE_WIDTH = 8
+        self.resetTimeout = 2000
+        self.previousStepTime = 0
+        self.minAnalogInputVoltage = 0.9
+        self.screen = 1
+        self.configMode = False
+        self.screenIndicator = ['.  ', ' . ', '  .' ]
+
+        self.MAX_DIVISION = 128
+        self.clockDivisions = []
+        for n in range(self.MAX_DIVISION+1):
+            self.clockDivisions.append(n)
+        
+        # When enabled, set msDriftCompensation to 30
+        # When disabled, set msDriftCompensation to 28
+        self.DEBUG = False
+
+        # In testing a 28ms drift was found at all tempos
+        # Adding this offset brings the BPM and pulse width back into a reasonable tolerance
+        self.msDriftCompensation = 28
+
+        # Vars to drive UI on screen3
+        self.markerPositions = [ [0, 1], [0, 11], [0, 21], [52, 1], [52, 11], [52, 21]]
+        self.activeOption = 1
+        self.previousselectedDivision = ''
+
+        # Get working vars
+        self.loadState()
+        self.calcSleepTime()
+        self.getPulseWidth()
+
+        # Get asyncio event loop object
+        self.el = asyncio.get_event_loop()
+        self.el.run_forever()
```

Comment:
[question] Is it necessary to call `run_forever()` in two places in this script? (the other being `MasterClock.main()` line 133.

Reply:
I tested it and only the one in the wrapper class is actually needed. I removed the one in the inner class.